### PR TITLE
✨ Add CreateIBAN method to create a virtual iban for a wallet

### PR DESCRIPTION
**Pull request by Sh0ri**
## Changes

Add new function `CreateVirtualIBAN` in order to create a virtual iban to a wallet

## Real case testing
- Clone lemonway repo, checkout this branch, launch `npm install`
- In your repo using `lemonway` :
  - change `package.json` to link to local lemonway package: `"lemonway": "file:./path/to/lemonway",`  
  - launch `npm install` to take this change into account

### CODE REVIEW

File: test/wallet/createIban.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    const lemonwayIban = await createdWallet.createIBAN(chance.ip(), {
+      wallet: createdWallet.id,
+      country: 'FR'
+    });
+
+    expect(lemonwayIban.id).to.not.be.undefined;
+
+    const infos = await lemonway.Wallet.getWalletDetails(chance.ip(), createdWallet);
+    console.log(infos.wallet.ibans);
```
Comment (jguepin) [inline test/wallet/createIban.js]: We don't need that 😉

File: test/wallet/createIban.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+const expect = require('chai').expect;
+const Chance = require('chance');
+const moment = require('moment');
+
+const Lemonway = require('../../');
+
+const chance = new Chance();
+
+describe.only('register iban', function () {
```
Comment (jguepin) [inline test/wallet/createIban.js]:
```suggestion
describe('register iban', function () {
```

File: lib/client.js [hunk 3]
```diff
@@ -318,6 +318,12 @@ Client.prototype.registerIBAN = function (ip, _opts) {
   return this._request('RegisterIBAN', ip, opts).get('IBAN');
 };
 
+Client.prototype.createIBAN = function(ip, _opts) {
```
[the remark this answers is not in the document]
Comment (jguepin) [reply lib/client.js]: I would have named that `createVirtualIBAN` to make a clear difference with the `registerIBAN` method.

**Review by achoisy — APPROVED** [review]
Tested in real case and working just fine 👍🏼 

Great work 🎖️

File: lib/client.js [hunk 4]
```diff
@@ -351,6 +351,15 @@ Client.prototype.registerIBAN = function (ip, _opts) {
   return this._request('RegisterIBAN', ip, opts).get('IBAN');
 };
 
+Client.prototype.createVirtualIBAN = function(ip, _opts) {
```
Comment (Jurollet) [inline lib/client.js:354]:
```suggestion
Client.prototype.createVirtualIban = function(ip, _opts) {
```

? to follow internal guidelines

File: lib/factories/wallet-factory.js [hunk 5]
```diff
@@ -321,6 +322,16 @@ function WalletFactory (lemonway) {
       });
   };
 
+  this.createVirtualIBAN = async function (ip, wallet, _opts) {
```
Comment (Jurollet) [inline lib/factories/wallet-factory.js:325]:
```suggestion
  this.createVirtualIban = async function (ip, wallet, _opts) {
```

File: lib/models/VirtualIBAN.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+function VirtualIBAN (data) {
```
Comment (Jurollet) [inline lib/models/VirtualIBAN.js:3]:
```suggestion
function VirtualIban (data) {
```

File: test/wallet/createVirtualIban.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
+const { expect } = require('chai');
+const Chance = require('chance');
+const moment = require('moment');
+
+const Lemonway = require('../../');
+
+const chance = new Chance();
+
+describe('createVirtualIban', function () {
+  this.timeout(2000000);
```
Comment (Jurollet) [inline test/wallet/createVirtualIban.js:10]: Why ?

File: test/wallet/createVirtualIban.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    const lemonwayIban = await wallet.createVirtualIBAN(chance.ip(), {
+      wallet: wallet.id,
+      country: 'FR'
+    });
+
+    expect(lemonwayIban).to.have.keys(['id', 'iban', 'bic', 'holder', 'domiciliation', 'status', 'maxAvailableIbanPerWalletLeft', 'maxAvailableIbanInTotalLeft']);
+
+    const updatedWallet = await lemonway.Wallet.getWalletDetails(chance.ip(), wallet);
+    expect(updatedWallet.wallet.ibans.length).to.eql(1);
+    expect(updatedWallet.wallet.ibans[0].iban).to.eql(lemonwayIban.iban);
```
Comment (Jurollet) [inline test/wallet/createVirtualIban.js:43]: Wow but ain't the ibans for performing money outs also stated in this array ?

How would we differentiate them in such case ?

**Review by Jurollet — CHANGES_REQUESTED** [review]

File: test/wallet/createVirtualIban.js (hunk 7 above)
Comment (Deprasos) [reply test/wallet/createVirtualIban.js:10]: All tests have this 🙈 Really don't know why 🤔 @jguepin any idea ?

File: lib/client.js (hunk 4 above)
Comment (Deprasos) [reply lib/client.js:354]: It was to be coherent with existing methods in the package `registerIBAN`, `createVCC`, `moneyInSDDInit`, `unregisterSDDMandate`, `registerSDDMandate` 🙈

File: test/wallet/createVirtualIban.js (hunk 8 above)
Comment (Deprasos) [reply test/wallet/createVirtualIban.js:43]: Nice catch. Indeed we are doing 
```
if (_.get(data, 'IBANS')) {
  this.ibans = _.map(_.get(data, 'IBANS.IBAN', _.get(data, 'IBANS')), function (iban) {
    return new IBAN(iban);
  });
}
```
I will check maybe we could create two arrays 👍

Comment (Deprasos) [reply test/wallet/createVirtualIban.js:43]: I've tested and when fetching the wallet details using `get` or `getWalletDetails`, lemonway returns all ibans in `IBANS`. 
```
IBANS LEMONWAY PACKAGE {
  IBAN: [
    {
      ID: '63494',
      S: '5',
      DATA: 'FR3530003000406794775792O03',
      SWIFT: 'FAKEBICXXX',
      HOLDER: 'FR Business'
    },
    {
      ID: '63495',
      S: '5',
      DATA: 'FR7630004029990001028202793',
      SWIFT: 'BNPAFRPPXXX',
      HOLDER: 'LEMONWAY'
    }
  ]
}
``` 
Only way to differentiate would be to use `HOLDER` but not sure we should do that. Let's have all ibans in the same array and let users to differentiate them when using the package.

File: lib/client.js (hunk 4 above)
Comment (Jurollet) [reply lib/client.js:354]: Ok Ok …

File: test/wallet/createVirtualIban.js (hunk 8 above)
Comment (Jurollet) [reply test/wallet/createVirtualIban.js:43]: That's pretty sad but we cannot do much about it indeed.

File: test/wallet/createVirtualIban.js (hunk 7 above)
Comment (jguepin) [reply test/wallet/createVirtualIban.js:10]: Lemon Way sandbox used to be slow to answer calls. We can try to remove/reduce it.

**Review by Sh0ri — COMMENTED** [review]
just in case you need help to test it in real case locally : https://stackoverflow.com/questions/52965414/how-to-use-local-package-rather-than-remote-package-in-nodejs

**Review by Jurollet — APPROVED** [review]
Tested real case OK

**Comment by Jurollet** [comment]
Nice work from outta space @Sh0ri ! Much love 🫶

**Pull request merged**